### PR TITLE
Fix caller transaction graph handoff

### PR DESCRIPTION
**Pull request by haiderz07**
## Summary

- track the top-level PostgreSQL transaction that owns caller-mode `df.start()` graph rows
- probe graph visibility and `pg_xact_status()` with bounded single-shot activities plus deterministic durable backoff
- retry transient database failures, distinguish whole-transaction aborts from committed savepoint rollbacks, and compact long waits with `continue_as_new`
- preserve historical replay by retaining the original activity name and raw instance-ID input for orchestration inputs without an origin xid

## Problem

`df.start()` persists `df.instances` / `df.nodes` in the caller transaction but commits the duroxide start separately. The graph loader previously stopped waiting after five seconds. A legal transaction held for eight seconds then committed with the control-plane row still `pending`, no SQL effect, and the engine execution terminally `Failed` with `Instance <id> not found after 5s`.

Synthetic reproducer:

```sql
BEGIN;
SELECT df.start('INSERT INTO effects VALUES (''{sys_instance_id}'')');
SELECT pg_sleep(8);
COMMIT;
```

The new regression also terminates a lock-blocked management backend and holds the lock long enough for the next probe to time out, proving both transient error paths retry and the effect still occurs exactly once.

## Compatibility

No extension schema, ABI, or persisted `df` data changes. Historical `FunctionInput` payloads default `origin_xid` to `None` and schedule `pg_durable::activity::load-function-graph` with the same raw input bytes and operation order. New starts alone use the versioned transaction-aware probe activity. Upgrade B1 passed against every supported v0.2.2-v0.2.6 schema on PostgreSQL 17 and 18.

## Validation

- `cargo fmt -p pg_durable -- --check`
- PG17 + PG18 `cargo build` and `cargo clippy -- -D warnings`
- PG17 + PG18 `cargo pgrx test`: 333 passed, 0 failed, 16 ignored each
- PG17 + PG18 full local E2E: 56 passed, 0 failed each
- new transaction regression: 3 consecutive PG17 runs, plus PG18
- PG17 + PG18 upgrade tests: 86 passed each
- PG17 + PG18 pg_regress: 6 passed each
- PG17 + PG18 shutdown and extension-epoch race regressions
- synthetic immediate-stop and worker-kill matrices across SQL, JOIN, RACE, loop, timer, signal, and cancellation

### CODE REVIEW

File: src/activities/load_function_graph.rs [hunk 1]
```diff
@@ -161,5 +158,281 @@ pub async fn execute(
...
+                                &ctx,
+                                "post-commit graph load",
+                                &input.instance_id,
+                                error,
+                            );
+                        }
+                        Err(LoadGraphError::Permanent(error)) => return Err(error),
+                    }
+                }
+                Ok(None) => TransactionGraphProbe::CommittedMissing,
```
[the remark this answers is not in the document]
Comment (haiderz07) [reply src/activities/load_function_graph.rs:650]: Fixed in `src/activities/load_function_graph.rs`. After `pg_xact_status()` reports \"committed\", we now query `pg_catalog.pg_visible_in_snapshot(\::text::xid8, pg_catalog.pg_current_snapshot())` before trusting a re-read. If the xid isn't snapshot-visible yet, we return `Retry` (see the new `CommittedProbeStep`/`committed_probe_step()` pure function). Only once the xid is confirmed snapshot-visible do we re-read and treat a missing graph as a genuine `CommittedMissing`. Added `committed_probe_step_awaits_snapshot_visibility_until_confirmed` unit test covering both branches of this boundary.

File: src/activities/load_function_graph.rs [hunk 2]
```diff
@@ -161,5 +158,281 @@ pub async fn execute(
...
+                "graph visibility check",
+                &input.instance_id,
+                "timed out after 2s",
+            );
+        }
+    };
+
+    match visible {
+        Ok(Some(row)) => {
+            let loaded = match tokio::time::timeout(
```
[the remark this answers is not in the document]
Comment (haiderz07) [reply src/activities/load_function_graph.rs:503]: Fixed in `src/activities/load_function_graph.rs`. Graph loading now has its own separate `GRAPH_LOAD_QUERY_TIMEOUT` (20s client-side), decoupled from the short `TRANSACTION_PROBE_QUERY_TIMEOUT` (2s) which is now reserved for cheap visibility/transaction-status probes only. `fetch_node_rows()` also applies Postgres-side `SET LOCAL statement_timeout` (15s) and `SET LOCAL lock_timeout` (5s) inside its own transaction, so a lock- or runaway-query condition fails fast with a proper SQLSTATE (`57014`/`55P03`) instead of relying solely on the client-side timeout. A stuck load now also counts against the new bounded retry budget described below rather than looping forever.

File: src/activities/load_function_graph.rs [hunk 3]
```diff
@@ -161,5 +158,281 @@ pub async fn execute(
...
+                Ok(graph) => serialize_probe(&TransactionGraphProbe::Ready { graph }),
+                Err(LoadGraphError::Retryable(error)) => {
+                    retry_probe(&ctx, "graph load", &input.instance_id, error)
+                }
+                Err(LoadGraphError::Permanent(error)) => Err(error),
+            };
+        }
+        Ok(None) => {}
+        Err(e) => {
+            return retry_probe(&ctx, "graph visibility check", &input.instance_id, e);
```
[the remark this answers is not in the document]
Comment (haiderz07) [reply src/activities/load_function_graph.rs]: Fixed in `src/activities/load_function_graph.rs`. Added `classify_sqlstate()`/`classify_sqlx_error()`: an explicit allowlist of genuinely transient SQLSTATEs (connection-exception class `08xxx`, `40001` serialization_failure, `40P01` deadlock_detected, our own `57014`/`55P03` statement/lock timeouts, and `57P01`/`57P02`/`57P03` admin-cancel/recovery conflicts) is treated as `Retry`; everything else -- including `42501`, `42883`, `42P01`, and any unrecognized code -- is classified `Permanent` and now fails the activity immediately with the SQLSTATE embedded via `classified_load_graph_error()`/`probe_error_outcome()`. Also added a bounded transient-retry budget (`MAX_GRAPH_RETRY_ATTEMPTS = 20`, tracked via the new `graph_retry_attempt` field, independent from the caller-transaction `InProgress` wait counter) in `execute_function_graph.rs`, so even allowlisted transient failures cannot poll forever. Unit tests cover the allowlist boundary (`classify_sqlstate_allows_connection_and_our_own_timeout_codes`, `classify_sqlstate_treats_privilege_and_schema_errors_as_permanent`, `classify_sqlstate_defaults_unknown_codes_to_permanent`) and the retry-budget boundary (`graph_retry_budget_exceeded_allows_exactly_max_attempts`).

File: tests/e2e/sql/68_long_caller_transaction.sql [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+    END IF;
+    IF NOT pg_terminate_backend(target_pid) THEN
+        RAISE EXCEPTION
+            'TEST FAILED [transient retry]: could not terminate graph-probe backend %',
+            target_pid;
+    END IF;
+
+    -- Leave time for the killed query to return Retry and the next lock-blocked
+    -- probe to hit its bounded query timeout before releasing the table lock.
+    PERFORM pg_sleep(3);
```
[the remark this answers is not in the document]
Comment (haiderz07) [reply tests/e2e/sql/68_long_caller_transaction.sql]: Rewrote the transient-retry scenario in `tests/e2e/sql/68_long_caller_transaction.sql`. The probe backend is now identified by its exact query text (`SELECT root_node, ... AS submitted_by ... FROM df.instances`), avoiding the worker's unrelated periodic retention-cleanup query which reads the same table and also blocks on the same lock. The test now asserts a strictly later `query_start` for a *second* attempt (proving the connection-failure retry path actually ran after `pg_terminate_backend`), and then a *third*, still-later `query_start` before releasing the lock (proving the independent, bounded client-side timeout-retry path also ran on its own, not just resumption after the lock is freed). We key on `state = 'active'` rather than `wait_event_type = 'Lock'`, since Postgres' periodic deadlock-check wakeup clears `wait_event_type` back to `NULL` for most of an actual lock wait even though the backend is still blocked. We also compare `query_start` rather than `pid`, since the worker's management connection pool may legitimately reuse the same physical backend across retries. While validating this against the harness we also found and fixed a real bug in the test itself: `pg_stat_activity`'s snapshot is cached for the lifetime of the enclosing transaction, so a long-lived `DO` block must call `pg_stat_clear_snapshot()` before every poll to observe live backend state -- without it, every iteration after the first silently re-read the same frozen snapshot. Validated with 8 back-to-back passes across PG17 and PG18 (5 on PG17, 3 on PG18) plus the full 54-test E2E suite on both majors and the 86-test upgrade suite.

File: tests/e2e/sql/68_long_caller_transaction.sql [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+    LOOP
+        SELECT i.status, i.output INTO engine_status, engine_output
+        FROM pg_temp.engine_info(v_instance_id) i;
+        EXIT WHEN lower(COALESCE(engine_status, '')) = 'failed' OR attempts >= 300;
+        PERFORM pg_sleep(0.1);
+        attempts := attempts + 1;
+    END LOOP;
+
+    IF lower(COALESCE(engine_status, '')) != 'failed'
```
[the remark this answers is not in the document]
Comment (haiderz07) [reply tests/e2e/sql/68_long_caller_transaction.sql:338]: Fixed in `tests/e2e/sql/68_long_caller_transaction.sql`. Added `engine_output IS NULL OR ...` to both the whole-rollback and savepoint-rollback assertions, so a missing diagnostic on a `failed` status now correctly fails the test instead of silently passing via PL/pgSQL's three-valued `false OR NULL = NULL` logic.

**Comment by haiderz07** [comment]
@microsoft-github-policy-service agree

**Comment by haiderz07** [comment]
Rebased this branch onto current \main\ (\3a8dbb\, #368) to resolve the \CHANGELOG.md\ merge conflict that appeared after the v0.2.7 release commit landed. Combined both edits: kept upstream's dated header/PR-number annotations and the new Changed section, and re-inserted the Caller-transaction handoff bullet. No other files touched by the rebase. New head: 80d3df0. Re-ran the full validation suite (fmt, clippy -D warnings on PG17+PG18, pgrx unit tests 309/0/16 on both majors, full E2E 56/0 on both majors including 68_long_caller_transaction, upgrade tests 86/86, plus 3 additional repeat runs of 68_long_caller_transaction on each major) against this rebased tree -- all green.

File: CHANGELOG.md [hunk 6]
```diff
@@ -16,6 +16,7 @@ Pre-1.0 note: while `pg_durable` is in major version `0`, minor releases may inc
 
 ### Fixed
 
+- **Caller-transaction handoff:** `df.start()` now tracks the originating transaction until it commits or aborts, so legal caller transactions lasting more than five seconds no longer leave a `pending` `df.instances` row paired with a failed engine execution. Graph admission uses durable backoff and bounded-history compaction rather than holding a worker connection while it waits.
```
[the remark this answers is not in the document]
Comment (pinodeca) [reply CHANGELOG.md]: Addressed. Rebased onto current `origin/main` (0.2.8 development cycle, #371 baseline `37a24f4`) and moved the caller-transaction handoff entry out of the already-published `0.2.7` section into the `0.2.8` Unreleased section. Pushed as `2a4478d`.

File: src/orchestrations/execute_function_graph.rs [hunk 7]
```diff
@@ -121,6 +122,155 @@ impl From<&str> for NodeError {
...
+                    ctx.trace_info(format!(
+                        "Compacting graph-admission history for instance {} after {} polls",
+                        input.instance_id, polls_in_execution
+                    ));
+                    return ctx.continue_as_new(continuation_json).await;
+                }
+            }
+            TransactionGraphProbe::Retry => {
+                retry_attempt = retry_attempt.saturating_add(1);
+                if graph_retry_budget_exceeded(retry_attempt) {
```
[the remark this answers is not in the document]
Comment (pinodeca) [reply src/orchestrations/execute_function_graph.rs:238]: Addressed in `2a4478d`. All terminal `df.instances` writes (`completed`/`failed`, including the graph-load-failure path) now go through `finalize_instance_status()`, which durably retries the `update-instance-status` activity across durable timers instead of a best-effort fire-and-forget `let _ = ...`. The retry budget is bounded (`MAX_STATUS_FINALIZE_ATTEMPTS = 20`), so a transient management-plane/pool outage self-heals once the database recovers, while a persistently unhealthy worker can't spin forever; on budget exhaustion the divergence is surfaced in the orchestration trace rather than silently hidden. Retries (activity results + timer fires) are recorded in duroxide history, so replay follows the same attempt sequence deterministically.

File: src/activities/load_function_graph.rs [hunk 8]
```diff
@@ -161,5 +276,422 @@ pub async fn execute(
...
+            input.origin_xid, input.instance_id
+        ));
+    }
+
+    ctx.trace_info(format!(
+        "Probing graph visibility for instance {} from origin transaction {}",
+        input.instance_id, input.origin_xid
+    ));
+
+    let visible = match tokio::time::timeout(
```
[the remark this answers is not in the document]
Comment (pinodeca) [reply src/activities/load_function_graph.rs:484]: Addressed in `2a4478d`. Every cheap probe now runs inside a short-lived transaction (`begin_probe_tx`) that applies server-side `SET LOCAL statement_timeout = '1500ms'` and `SET LOCAL lock_timeout = '1500ms'`, set *below* the 2s client-side `TRANSACTION_PROBE_QUERY_TIMEOUT`. This covers the visibility check (`find_visible_instance`), the transaction-status probe (`probe_origin_transaction_status`), and the post-commit snapshot check (`probe_snapshot_visible`); the role/superuser probe (`is_role_superuser_name` in `types.rs`) gets the same treatment. PostgreSQL now cancels a blocked probe with `57014`/`55P03` and returns the connection cleanly to the pool before the client-side `tokio::time::timeout` would drop the in-flight future and force the pool to drain/discard it — so a conflicting lock on `df.instances` can no longer exhaust the small management pool through successive durable retries. `SET LOCAL` is transaction-scoped and discarded on the (read-only) rollback, so no modified timeout leaks onto a pooled connection reused by unrelated work.

**Review by pinodeca — APPROVED** [review]
